File: Leetcode 417 - pacific-atlantic-water-flow/417.pacific-atlantic-water-flow.cpp
/*
 * @lc app=leetcode id=417 lang=cpp
 *
 * [417] Pacific Atlantic Water Flow
 */

// @lc code=start
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<int>> pacificAtlantic(vector<vector<int>> &heights) {
    vector<vector<int>> res;
    int m = heights.size(), n = heights[0].size();
    vector<vector<bool>> pacific(m, vector<bool>(n, false));
    vector<vector<bool>> atlantic(m, vector<bool>(n, false));

    for (int i = 0; i < m; i++) {
      dfs(i, 0, heights, pacific, -1);
      dfs(i, n - 1, heights, atlantic, -1);
    }
    for (int j = 0; j < n; j++) {
      dfs(0, j, heights, pacific, -1);
      dfs(m - 1, j, heights, atlantic, -1);
    }

    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (pacific[i][j] && atlantic[i][j]) {
          res.push_back({i, j});
        }
      }
    }

    return res;
  }

  void dfs(int i, int j, vector<vector<int>> &heights,
           vector<vector<bool>> &visited, int prev) {
    if (i < 0 || j < 0 || i >= heights.size() || j >= heights[0].size() ||
        visited[i][j] || heights[i][j] < prev) {
      return;
    }

    visited[i][j] = true;

    dfs(i + 1, j, heights, visited, heights[i][j]);
    dfs(i - 1, j, heights, visited, heights[i][j]);
    dfs(i, j + 1, heights, visited, heights[i][j]);
    dfs(i, j - 1, heights, visited, heights[i][j]);
  }
};

int main() {
  Solution sol;

  vector<vector<int>> heights = {{1, 2, 2, 3, 5},
                                 {3, 2, 3, 4, 4},
                                 {2, 4, 5, 3, 1},
                                 {6, 7, 1, 4, 5},
                                 {5, 1, 1, 2, 4}};
  vector<vector<int>> result = sol.pacificAtlantic(heights);
  for (auto &row : result) {
    for (auto &col : row) {
      cout << col << " ";
    }
    cout << endl;
  }
  // {0, 4}, {1, 3}, {1, 4}, {2, 2}, {3, 0}, {3, 1}, {4, 0}

  //   heights = {{1}};
  //   result = sol.pacificAtlantic(heights);
  //   for (auto &row : result) {
  //     for (auto &col : row) {
  //       cout << col << " ";
  //     }
  //     cout << endl;
  //   }
  // {0, 0}
}
// @lc code=end

File: Leetcode 417 - pacific-atlantic-water-flow/417.pacific-atlantic-water-flow.java
/*
 * @lc app=leetcode id=417 lang=java
 *
 * [417] Pacific Atlantic Water Flow
 */

// @lc code=start

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {

  private int[][] directs = { { -1, 0 }, { 0, 1 }, { 1, 0 }, { 0, -1 } };

  public List<List<Integer>> pacificAtlantic(int[][] heights) {
    List<List<Integer>> result = new ArrayList<>();
    int rows = heights.length, cols = heights[0].length;
    boolean[][] pac = new boolean[rows][cols];
    boolean[][] atl = new boolean[rows][cols];

    for (int i = 0; i < rows; i++) {
      // start from the first column, and check if the cell is reachable from the pacific
      dfs(heights, i, 0, pac, -1);
      // start from the last column, and check if the cell is reachable from the atlantic
      dfs(heights, i, cols - 1, atl, -1);
    }

    for (int j = 0; j < cols; j++) {
      // start from the first row, and check if the cell is reachable from the pacific
      dfs(heights, 0, j, pac, -1);
      // start from the last row, and check if the cell is reachable from the atlantic
      dfs(heights, rows - 1, j, atl, -1);
    }

    // check if the cell is reachable from both the pacific and atlantic
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        if (pac[i][j] && atl[i][j]) {
          result.add(Arrays.asList(i, j));
        }
      }
    }

    return result;
  }

  // Time Complexity: O(m * n)
  // Space Complexity: O(m * n)

  private void dfs(
    int[][] heights,
    int i,
    int j,
    boolean[][] visited,
    int prev
  ) {
    if (
      i < 0 ||
      j < 0 ||
      i >= heights.length ||
      j >= heights[0].length ||
      visited[i][j] ||
      heights[i][j] < prev
    ) {
      return;
    }

    visited[i][j] = true;

    for (int[] direct : directs) {
      dfs(heights, i + direct[0], j + direct[1], visited, heights[i][j]);
    }
  }
}
// @lc code=end
